addons/portal/models/mail_thread.py

```diff
@@ -4,7 +4,70 @@
 import hashlib
 import hmac
 
-from odoo import fields, models, _
+from functools import wraps
+from inspect import Parameter, signature
+
+from odoo import fields, models, api, _
+from odoo.http import request
+from odoo.addons.bus.websocket import wsrequest
+from odoo.addons.portal.controllers.mail import _check_special_access
+
+
+def check_portal_access(func):
+    """Decorate a function to extract the portal token from the request.
+    The portal token is then available on the context of the current
+    request.
+    """
+
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+        req = request or wsrequest
+        context = dict(req.env.context)
+        context.pop("portal_token", None)
+        portal_token = kwargs.pop("portal_token", None)
+        portal_hash = kwargs.pop("portal_hash", None)
+        portal_pid = kwargs.pop("portal_pid", None)
+        portal_res_id = kwargs.pop("portal_res_id", None) or kwargs.get("thread_id")
+        portal_res_model = kwargs.pop("portal_res_model", None) or kwargs.get(
+            "thread_model"
+        )
+        if portal_token or (portal_hash and portal_pid):
+            record = request.env[portal_res_model].browse(portal_res_id).sudo()
+            has_access = _check_special_access(
+                portal_res_model, portal_res_id, portal_token, portal_hash, portal_pid
+            )
+            if has_access:
+                req.update_context(
+                    portal_token=portal_token,
+                    portal_hash=portal_hash,
+                    portal_pid=portal_pid,
+                )
+                if req.env.user._is_public() and hasattr(record, "partner_id"):
+                    req.update_context(portal_partner=record.partner_id)
+        return func(self, *args, **kwargs)
+
+    old_sig = signature(wrapper)
+    params = list(old_sig.parameters.values())
+    new_param_index = next(
+        (
+            index
+            for index, param in enumerate(params)
+            if param.kind in [Parameter.VAR_POSITIONAL, Parameter.VAR_KEYWORD]
+        ),
+        len(params),
+    )
+    for param in [
+        "portal_token",
+        "portal_hash",
+        "portal_pid",
+        "portal_res_id",
+        "portal_res_model",
+    ]:
+        new_param = Parameter(param, Parameter.POSITIONAL_OR_KEYWORD, default=None)
+        params.insert(new_param_index, new_param)
+        new_param_index += 1
+    wrapper.__signature__ = old_sig.replace(parameters=params)
```

Comment:
If you check the latest version of `add_guest_to_context`, we no longer do this kind of change, simply updating a context value instead.

It would be better if this was simplified here.